fake.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package testing
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+)
+
+// FakeCallArgs holds all the args passed in a call.
+type FakeCallArgs map[string]interface{}
+
+// FakeCall records the name of a called function and the passed args.
+type FakeCall struct {
+	// Receiver is the fake for which the function was called. It is
+	// not required, particularly if the function is not a method.
+	Receiver interface{}
+
+	// Funcname is the name of the function that was called.
+	FuncName string
+
+	// Args is the set of arguments to the function.
+	Args FakeCallArgs
+}
+
+// Fake is used in testing to stand in for some other value, to record
+// all calls to faked methods/functions, and to allow users to set the
+// values that are returned from those calls. Fake is intended to be
+// embedded in another struct that will define the methods to track:
+//
+//    type fakeConn struct {
+//        *testing.Fake
```

Comment:
It needs to be a pointer so that the embedded fake can be shared between multiple fake types.  Being able to do so was actually super useful in tests I've written.